### ISSUE TITLE
[TW-1581] Deleting a task with dependencies on other tasks doesn't show new urgency on its depenencies on the first listing afterwards

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1607]
_Ulf Eliasson on 2015-03-23T22:25:40Z says:_ 

 When deleting or setting a task to done that is depending on one or more tasks, the new urgency is not shown for the dependencies on the first listing of the tasks afterwards, with e.g. `task next` or `task list`. 
On executing `task next` or `task list` a second time the correct urgency is shown.

Tested on 2.4.2 and 2.4.3 from git.

To reproduce with an empty task list:
{code:none}
task add Task 1 && task add depend:1 Task 2 && yes | task 2 delete && task && task
{code}

Result:
{code:none}
[...]
[task next]

ID Description Urg 
 1 Task 1         8

1 task

[task next]

ID Description Urg 
 1 Task 1         0

1 task
{code}

Expected:
{code:none}
[...]
[task next]

ID Description Urg 
 1 Task 1         0

1 task

[task next]

ID Description Urg 
 1 Task 1         0

1 task
{code}

Screencast demonstrating the bug: https://asciinema.org/a/17960

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2015-03-23T22:25:40Z
Modified: 2015-03-28T09:51:26Z
```

**Comment by taskwarrior** [comment]
_Ulf Eliasson on 2015-03-25T14:24:40Z says:_ 

 Added test case demonstrating the bug.

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-03-26T18:01:19Z says:_ 

 Fixed in [d2e6c90446d231be955cd5c60137131f6034e049|https://git.tasktools.org/projects/TM/repos/task/commits/d2e6c90446d231be955cd5c60137131f6034e049].

Thank you for writing a test, since I would probably have been too lazy to get that done anytime soon (read: ever). In that light it is probably a bit too much to ask you to change it a tiny bit, but here we go:

Your current test checks for a particular value of urgency.blocking.coefficient, when it might be a better approach to check that running "task list", or better yet "task export" twice results in the exact same output. That way the test won't have to be changed in case the coefficient or the list report changes, and it might actually catch more bugs in case someone messes with other attributes.

So: If you change the test's behavior, or tell me to stop being lazy and to do it myself, I'll add it.

Either way, thanks for reporting this!

**Comment by taskwarrior** [comment]
_Ulf Eliasson on 2015-03-26T18:39:35Z says:_ 

 That was my initial attempt, to just compare that the output was equal between two listings.

The problem was that in some cases, maybe one in ten, that I executed the test case enough time passed between the two list that it printed 1s for the age and therefore it reported it as an error.

But you are right that the test case is not very robust. I tried a little bit by setting the coefficient in the setup, but it will not work if the output changes in some other way, the configuration parameter is renamed or something similar.

Maybe there is a better way to test it, or one could possibly write something that can check the output and ignoring the age, or a custom report that do not display the age.

My taskwarrior knowledge is limited but if you have an idea I can try it and change the test case, or I can just compare that they are equal but I do not like test-cases that give sporadic false positives :)

It would be interesting to make similar testcases for other cases because I have gotten the feeling sometimes that something I did didn't give the expected result, but then if I list again it seemed to have fixed itself. However, I didn't research it properly except for this case that was so obvious :)

{code}
not ok 1 - deleting task with dependencies, blocker should have 0 urgency
1.  FAIL: AssertionError on file ./tw-1581.t line 49 in test_urgency_after_delete: 'self.check_urgency_after("delete")':
1.        '
1.  ID D Description Urg
1.  -- - ----------- ----
1.   1   Task 1         0
1.  
1.  1 task
1.  ' != '
1.  ID Age D Description Urg
1.  -- --- - ----------- ----
1.   1 1s    Task 1         0
1.  
1.  1 task
1.  '
{code}

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-03-26T18:47:13Z says:_ 

 That is precisely why I mentioned "task export", I probably should have made that more clear.

With export you get (almost) raw data, so you can just check for equality. Not depending on the format of a certain report is a good idea anyway, because you want to make your test as robust as possible.

**Comment by taskwarrior** [comment]
_Ulf Eliasson on 2015-03-26T18:55:29Z says:_ 

 I read too fast, my eyes (or brain) skipped the "task export" bit completely.

I will update the test case.

**Comment by taskwarrior** [comment]
_Ulf Eliasson on 2015-03-26T19:00:46Z says:_ 

 That actually always fails, as the second task have id:2 in the first listing but id:0 in the second, another bug report?

I guess it is caused by the same thing, that gc runs after the export? But it feels like it should run before export as it does before other things that list tasks.

{code}
not ok 2 - setting task with dependencies as done, blocker should have 0 urgency
1.  FAIL: AssertionError on file ./tw-1581.t line 54 in test_urgency_after_done: 'self.check_urgency_after("done")':
1.        '{"id":1,"description":"Task 1","entry":"20150326T185959Z","modified":"20150326T185959Z","status":"pending","uuid":"046b5a97-7306-4841-b1d1-d8a22b5644d0","urgency":0}
1.  {"id":2,"depends":"046b5a97-7306-4841-b1d1-d8a22b5644d0","description":"Task 2","end":"20150326T185959Z","entry":"20150326T185959Z","modified":"20150326T185959Z","status":"completed","uuid":"eba73420-f8d0-4857-adfb-360820b5855b","urgency":0}
1.  ' != '{"id":1,"description":"Task 1","entry":"20150326T185959Z","modified":"20150326T185959Z","status":"pending","uuid":"046b5a97-7306-4841-b1d1-d8a22b5644d0","urgency":0}
1.  {"id":0,"depends":"046b5a97-7306-4841-b1d1-d8a22b5644d0","description":"Task 2","end":"20150326T185959Z","entry":"20150326T185959Z","modified":"20150326T185959Z","status":"completed","uuid":"eba73420-f8d0-4857-adfb-360820b5855b","urgency":0}
1.  '
{code}

**Comment by taskwarrior** [comment]
_Wim Schuermann on 2015-03-26T19:11:52Z says:_ 

 Hah, when I said "and it might actually catch more bugs in case someone messes with other attributes" I wasn't thinking of already existing bugs, but yes, that warrants another bug report. It is different from what is happening here, and TW-1582 is yet another incantation of a similar bug I noticed when testing whether I had fixed TW-1581.

GC is run for export, and more generally for all reports that display IDs. Since IDs should be valid (unless they're 0, which is special), and "2" is not valid after "task export" exits, yes, this is definitely a bug.
